Write item aggregate to MDC

jcocchi opened the pull request:
## All Submissions:

* [x] (Required) Have you followed the guidelines in our Contributing [document](./CONTRIBUTING.md)?
* [x] (Required) Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] I have updated the documentation accordingly.
* [ ] I have added tests to cover my changes. 
* [ ] All new and existing tests passed.
* [x] My code follows the code style of this project.
* [ ] I ran lint checks locally prior to submission.
* [x] (Required) Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

Issue Number: closes #21

## What is the new behavior?
Writes the item to MDC after the aggregation is done.

## Does this introduce a breaking change?
- [NO]


## Any relevant logs, error output, etc?
Will log any errors inserting to Cosmos.

ianphil commented on ch03/orchestrations/InventoryOrchestrator.cs:
```diff
@@ -0,0 +1,29 @@
+            MDC mdcItem = new MDC(incomingEvent.StoreId, incomingEvent.DivisionId, item);
+
+            // Write the item's aggregation to Cosmos DB
+            await context.CallActivityAsync("WriteToCosmos", mdcItem);
```
If I'm reading this correctly, this writes to cosmos after every aggregation?

davidkpiano replied:
Yes

codepossible replied:
I am still fuzzy about the decision to use the orchestrator to write to the Cosmos DB, instead of the entity as flow. Can you help me understand that flow. I know there is separation of concern and SRP is play here, but aren't we pushing too much responsibility to the Orchestrator here.

jcocchi replied:
We have a new Activity Function that writes to Cosmos, the orchestrator just ... orchestrates that call. 

We considered making a `WriteToCosmos` function in the entity itself instead of as a separate activity function, but entities behave differently than regular classes and I couldn't figure out how to create a static `CosmosClient` in the constructor of an entity (I actually don't see any entity constructor at all, maybe @brandonh-msft has an idea of how to do this?)

ianphil replied:
I was on the fence about this also, but as I thought it though (when we need to add temporal operations down the line) this pattern seems to emerge. I do think you could keep it in the entity for now using DI similar to what you are already doing with the activity: https://docs.microsoft.com/en-us/azure/azure-functions/durable/durable-functions-dotnet-entities#dependency-injection-in-entity-classes

jcocchi replied:
Brandon made a good point about DI being advanced for this open hack, I just refactored the activity function to use a static cosmos client instead of injecting the service. 
@ianphil - Do you think it would be worth it to refactor again to keep in DI and have the write in the entity for the purpose of this open hack? (totally fine doing the work, just not sure if we want to introduce the DI concept or not)

codepossible replied:
When I started think about it, the coaches repo and participant repo, were never intended to be the same. While reading this, take it with a pinch of salt, as this is my first foray into building an OpenHack, while @jcocchi , @brandonh-msft and @ianphil have more more experience with it.

I thought the coaches repo, as our attempt to solve the problem as "participants", so that we can build documentation which participants can use and figure out what to put in place, so they are able to complete the challenges.

We could freely experiment in the coach's repository, without ever having to share the solution out with participants, who will have their own skill level. It also gives the coaches to cross pollinate. 

Such as `ConfigureAwait(false)` call which I think, no one on this team will ever forget to add on C# code.... :)

brandonh-msft commented on ch03/CosmosWriterActivity.cs:
```diff
@@ -0,0 +1,29 @@
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Spike.Inventory
```
were we going to be changing these namespaces?

jcocchi replied:
I changed them all to be uniform but this is what I landed on. It's probably not the best name though, what do you suggest?

brandonh-msft replied:
`InventoryManagement` is best gut reaction right now ;) Just didn't think `Spike` was appropriate here.

codepossible replied:
OMG, I made this change on namespace, without looking into this PR... :)

jcocchi replied:
I didn't end up changing the namespaces here but we can always do it in another PR if needed.

brandonh-msft commented on ch03/CosmosWriterActivity.cs:
```diff
@@ -0,0 +1,29 @@
+{
+    class CosmosWriterActivity
+    {
+        private static ICosmosService _cosmosService;
```
we usually see `readonly` here as well ( `private readonly` ) so can't be reassigned outside constructor. I don't think this should be static since the class isn't static

jcocchi replied:
good point on the `readonly`, I'll update that! It's a best practice to use a static Cosmos client for Azure Functions, maybe I'll make the class static as well to match

brandonh-msft replied:
Cosmos _client_ yes. But this is your own service - it should, therefore, take care of the client being static. I think it'll ultimately go away when you nix DI in which case, yup, `private static readonly CosmosClient` would be the way to go

jcocchi replied:
Done, static client is created in my own service and this activity only calls that static method now

brandonh-msft commented on ch03/CosmosWriterActivity.cs:
```diff
@@ -0,0 +1,29 @@
+        [FunctionName("WriteToCosmos")]
+        public async Task WriteToCosmosAsync([ActivityTrigger] MDC item)
+        {
+            await _cosmosService.WriteItemToMDC(item);
```
don't await - just return the task.

jcocchi replied:
done

brandonh-msft commented on ch03/Readme.md:
```diff
@@ -0,0 +1,69 @@
+# CSharp Durable Functions Counter Starter Template
```
I think you want to change this heading - yeah?

jcocchi replied:
done

brandonh-msft commented on ch03/Readme.md:
```diff
@@ -0,0 +1,69 @@
+
+## Run
+
+To run the project locally enter `func start` at the CLI, or in Visual Studio/VS Code, press `<F5>` in the `ch02` workspace. You should see the function url `http://localhost:7071/InventoryOrchestrator_HttpStart`. Use Postman (or some other tool) to submit a `POST` request with the sample body structure below. You should see a document either written (if it is the first message for this product i ...
```
ch03 workspace, right?

jcocchi replied:
done

brandonh-msft commented on ch03/host.json:
```diff
@@ -0,0 +1,19 @@
+            }
+        }
+    },
+    "extensions": {
```
see my comments in Teams on this; for OpenHacks I don't believe we should deviate from defaults.

jcocchi replied:
I have no opinion on this so I'm okay changing it, but I didn't write this so want to make sure @davidkpiano is okay with removing it too

brandonh-msft commented on ch03/models/InventoryEvent.cs:
```diff
@@ -0,0 +1,44 @@
+        [JsonProperty("description")]
+        public string Description { get; set; }
+
+        [JsonProperty(PropertyName = "lastShipmentTimestamp", NullValueHandling = NullValueHandling.Ignore)]
```
I think you need to sync w/ @gavination 's implementation on this

jcocchi replied:
I had this originally and he changed it when he did the merge, I can change mine to match his (this is a "changed in one folder after copied to the other folder issue" seems to be popping up a lot)

brandonh-msft replied:
100% agree

jcocchi replied:
I was getting build errors using @gavination's version with a nullable DateTime when assigning values to that property so I ended up keeping my implementation everywhere instead.

brandonh-msft commented on ch03/orchestrations/InventoryOrchestrator.cs:
```diff
@@ -0,0 +1,29 @@
+            var entityId = new EntityId("StoreEntity", incomingEvent.StoreId);
+
+            // Update item based on previous state and get new state back
+            var item = await context.CallEntityAsync<Item>(entityId, "Aggregate", incomingEvent);
```
`.ConfigureAwait(false)`

jcocchi replied:
For whatever reason when I use `.ConfigureAwait(false)` on this line I get this error. From the error message it looks like it might actually be erroring on line 27 for the next await. I'm wondering if it's because the `Aggregate` function is not actually async but we had to add `Task.FromResult()` because entity functions can't return values unless they are async. Any ideas?

```
[5/23/2020 12:03:47 AM] d9d2ffe2c47c45179e13b14ac31d43b9: Function 'InventoryOrchestrator (Orchestrator)' failed with an error. Reason: System.InvalidOperationException: Multithreaded execution was detected. This can happen if the orchestrator function code awaits on a task that was not created by a DurableOrchestrationContext method. More details can be found in this article https://docs.microsoft.com/en-us/azure/azure-functions/durable-functions-checkpointing-and-replay#orchestrator-code-constraints.
[5/23/2020 12:03:47 AM]    at Microsoft.Azure.WebJobs.Extensions.DurableTask.DurableOrchestrationContext.ThrowIfInvalidAccess() in d:\a\r1\a\azure-functions-durable-extension\src\WebJobs.Extensions.DurableTask\ContextImplementations\DurableOrchestrationContext.cs:line 1046
[5/23/2020 12:03:47 AM]    at Microsoft.Azure.WebJobs.Extensions.DurableTask.DurableOrchestrationContext.CallDurableTaskFunctionAsync[TResult](String functionName, FunctionType functionType, Boolean oneWay, String instanceId, String operation, RetryOptions retryOptions, Object input, Nullable`1 scheduledTimeUtc) in d:\a\r1\a\azure-functions-durable-extension\src\WebJobs.Extensions.DurableTask\ContextImplementations\DurableOrchestrationContext.cs:line 442
[5/23/2020 12:03:47 AM]    at Spike.Inventory.InventoryOrchestrator.RunOrchestrator(IDurableOrchestrationContext context, ILogger logger) in C:\code\openhack-durable-entities-coach\ch03\orchestrations\InventoryOrchestrator.cs:line 27
[5/23/2020 12:03:47 AM]    at Microsoft.Azure.WebJobs.Host.Executors.VoidTaskMethodInvoker`2.InvokeAsync(TReflected instance, Object[] arguments) in C:\projects\azure-webjobs-sdk-rqm4t\src\Microsoft.Azure.WebJobs.Host\Executors\VoidTaskMethodInvoker.cs:line 21
[5/23/2020 12:03:47 AM]    at Microsoft.Azure.WebJobs.Host.Executors.FunctionInvoker`2.InvokeAsync(Object instance, Object[] arguments) in C:\projects\azure-webjobs-sdk-rqm4t\src\Microsoft.Azure.WebJobs.Host\Executors\FunctionInvoker.cs:line 52
[5/23/2020 12:03:47 AM]    at Microsoft.Azure.WebJobs.Extensions.DurableTask.TaskOrchestrationShim.Execute(OrchestrationContext innerContext, String serializedInput) in d:\a\r1\a\azure-functions-durable-extension\src\WebJobs.Extensions.DurableTask\Listener\TaskOrchestrationShim.cs:line 96. IsReplay: False. State: Failed. HubName: SpikeHub. AppName: . SlotName: . ExtensionVersion: 2.2.1. SequenceNumber: 12.
```

jcocchi replied:
The configure awaits work on the other function calls, it's just this one that seems to be causing trouble

brandonh-msft replied:
Right, only **Activity** functions can make async calls. You need to `StartActivityAsync` (or whatever the API is) and make your `CallEntityAsync` call in there.

brandonh-msft replied:
hm. This works fine for me:
![image](https://user-images.githubusercontent.com/20270743/82738299-81f84300-9ceb-11ea-800e-1472e6af6773.png)

brandonh-msft commented on ch03/orchestrations/InventoryOrchestrator.cs:
```diff
@@ -0,0 +1,29 @@
+            MDC mdcItem = new MDC(incomingEvent.StoreId, incomingEvent.DivisionId, item);
+
+            // Write the item's aggregation to Cosmos DB
+            await context.CallActivityAsync("WriteToCosmos", mdcItem);
```
`.ConfigureAwait(false)`

jcocchi replied:
done

brandonh-msft commented on ch03/orchestrations/Startup.cs:
```diff
@@ -0,0 +1,25 @@
+﻿using Microsoft.Azure.Functions.Extensions.DependencyInjection;
```
oy. I'd strongly recommend against introducing Functions DI in an OpenHack.

jcocchi replied:
Yeah it is a bit advanced... this is just the coaches solution though so attendees can solve any way they see fit. Do you think it's okay to leave here or should we refactor it?

brandonh-msft replied:
ultimately, though, the coach needs to be able to guide attendees and/or proof the solution - I'd _highly_ doubt any (significant number) of the attendees start fabbing up Functions DI ;)

jcocchi replied:
Refactored DI out of this example

brandonh-msft commented on ch03/services/CosmosService.cs:
```diff
@@ -0,0 +1,47 @@
+
+            try
+            {
+                await container.UpsertItemAsync(item, new PartitionKey(item.Id));
```
`.ConfigureAwait(false)`

jcocchi replied:
done

jcocchi commented:
Okay I addressed most of the comments and I also tested end to end writing documents to the `onHand` and `shipments` container and making sure the appropriate updates end up in `mdc`.

Take another look when you get the chance @brandonh-msft  @ianphil 

Thanks!

ianphil commented on ch03/clients/OnHandUpdateFunc.cs:
```diff
@@ -0,0 +1,60 @@
+            databaseName: "inventory",
+            collectionName: "onHand",
+            ConnectionStringSetting = "CosmosDBConnectionString",
+            LeaseCollectionName = "leases")]IReadOnlyList<Document>
```
```suggestion
            LeaseCollectionName = "leases",
            CreateLeaseCollectionIfNotExists = true)]IReadOnlyList<Document>
```

ianphil commented on ch03/clients/ShipmentUpdateFunc.cs:
```diff
@@ -0,0 +1,59 @@
+            databaseName: "inventory",
+            collectionName: "shipments",
+            ConnectionStringSetting = "CosmosDBConnectionString",
+            LeaseCollectionName = "leases")]IReadOnlyList<Document> input,
```
```suggestion
            LeaseCollectionName = "leases",
            CreateLeaseCollectionIfNotExists = true)]IReadOnlyList<Document>
```

ianphil commented on ch03/entities/StoreEntity.cs:
```diff
@@ -0,0 +1,74 @@
+        
+        private void OnHandUpdate(InventoryEvent incomingEvent, Item currentItem)
+        {
+            currentItem.InventoryCount = incomingEvent.CountAdjustment;
```
`OnHand` event (`CountAdjustment `== 1000) arrives first `InventoryCount` == 1000
`Shipment` event (`CountAdjustment `== 3000) arrives second, `InventoryCount` == 4000 ✔️ 

`Shipment` event (`CountAdjustment `== 3000) arrives first, `InventoryCount` == 3000
`OnHand` event (`CountAdjustment `== 1000) arrives second `InventoryCount` == 1000  ❌

ianphil replied:
@davidkpiano @jcocchi looks like a great use for a state machine...
